### PR TITLE
feat(compiler): introduce pre-computed type system input

### PR DESCRIPTION
**Pull request by SimonSapin**
This allows extracting a parsed type system from Salsa
into read-only data that can be shared across threads,
with each thread using it as input to new compiler instances.

Fixes https://github.com/apollographql/apollo-rs/issues/406

### CODE REVIEW

**Comment by SimonSapin** [comment]
This PR is draft while I try it out in apollo-router

**Comment by SimonSapin** [comment]
Alright, I have a local router branch when this basically works as expected. We might want to add more precomputed things, but that can be further PRs.

**Comment by SimonSapin** [comment]
I looked through the diff again, this adds public APIs but afaict does not make breaking changes to existing ones

File: crates/apollo-compiler/src/database/hir_db.rs [hunk 1]
```diff
@@ -67,6 +68,9 @@ pub trait HirDatabase: InputDatabase + AstDatabase {
         name: String,
     ) -> Option<Arc<OperationDefinition>>;
 
+    /// Return an operation definition without a name corresponding to the file id.
+    fn find_unnamed_operation(&self, file_id: FileId) -> Option<Arc<OperationDefinition>>;
```
Comment (SimonSapin) [inline crates/apollo-compiler/src/database/hir_db.rs:72]: This is a drive-by new feature. Should it be called `find_anonymous_operation` instead? The spec uses both terms

File: crates/apollo-compiler/src/lib.rs [hunk 2]
```diff
@@ -79,6 +80,17 @@ impl ApolloCompiler {
...
+    pub fn set_precomputed_schema(&mut self, schema: Arc<PrecomputedTypeSystem>) {
+        if !self.db.type_definition_files().is_empty() {
+            panic!(
+                "Having both string inputs and pre-computed inputs \
+                 for type system definitions is not supported"
+            )
+        }
+        self.db.set_precomputed_input(Some(schema))
+    }
+
```
[the remark this answers is not in the document]
Comment (lrlna) [reply crates/apollo-compiler/src/lib.rs:93]: i am also wondering about the wording of `precomputed_schema`, should we just call this `type_system`, so the method would be `insert_type_system`?

**Review by lrlna — APPROVED** [review]
a few comments on naming things (oh no). let me know what you think?

File: crates/apollo-compiler/src/database/hir_db.rs (hunk 1 above)
Comment (lrlna) [reply crates/apollo-compiler/src/database/hir_db.rs:72]: jinx! i like `anonymous` better

File: crates/apollo-compiler/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (goto-bus-stop) [reply crates/apollo-compiler/src/lib.rs:93]: we have `create_schema`/`update_schema` right now for type definition files, as I think most people casually refer to the type system as the schema(?). i think this should use the same name as those

[the remark this answers is not in the document]
Comment (lrlna) [reply crates/apollo-compiler/src/lib.rs:93]: this api doesn't insert a full schema, but the whole type system, which is what i think the noun should reflect

[the remark this answers is not in the document]
Comment (SimonSapin) [reply crates/apollo-compiler/src/lib.rs:93]: The doc-comments for `create_*` use "Add", `add_*` would be good for those IMO and work well with `update_*`.

The reasoning for the separate `set_` naming convention is that it works differently:

* `add_` / `create_` assigns a new ID and returns it
* `update_` takes an ID as parameter
* `set_`: IDs are not involved. There’s just one precomputed thing, `set_` add it if there wasn’t one or updates it if there was

[the remark this answers is not in the document]
Comment (SimonSapin) [reply crates/apollo-compiler/src/lib.rs:93]: Proposal: https://github.com/apollographql/apollo-rs/pull/412

[the remark this answers is not in the document]
Comment (lrlna) [reply crates/apollo-compiler/src/lib.rs:93]: ok ok sold, let's do `add_`, `update_` and `set_`. 

i think this should be `set_type_system` though. the argument type is `type_system` after all, and it is very specifically different from a schema (has inputs, subtype_map + other things in the future likely?).

File: crates/apollo-compiler/src/database/hir.rs [hunk 3]
```diff
@@ -22,6 +25,18 @@ pub struct TypeSystemDefinitions {
     pub directives: ByName<DirectiveDefinition>,
 }
 
+/// Contains a `TypeSystemDefinition` together with:
+///
+/// * Other data that can be derived from it.
+/// * Relevant inputs
+#[derive(PartialEq, Eq, Debug)]
+pub struct PrecomputedTypeSystem {
```
Comment (lrlna) [inline crates/apollo-compiler/src/database/hir.rs:33]: If we keep this as `Precomputed`, by extension all other slightly computed types (subtype_map) should be prefixed with precomputed, and that's a bit wordy
```suggestion
pub struct TypeSystem {
```

File: crates/apollo-compiler/src/database/hir.rs [hunk 4]
```diff
@@ -22,6 +25,18 @@ pub struct TypeSystemDefinitions {
     pub directives: ByName<DirectiveDefinition>,
 }
 
+/// Contains a `TypeSystemDefinition` together with:
```
Comment (lrlna) [inline crates/apollo-compiler/src/database/hir.rs:28]:
```suggestion
/// Contains `TypeSystemDefinitions` together with:
```

File: crates/apollo-compiler/src/database/hir_db.rs [hunk 5]
```diff
@@ -23,6 +22,8 @@ pub trait HirDatabase: InputDatabase + AstDatabase {
     /// Return all type system definitions defined in the compiler.
     fn type_system_definitions(&self) -> Arc<TypeSystemDefinitions>;
 
+    fn precomputed_type_system(&self) -> Arc<PrecomputedTypeSystem>;
```
Comment (lrlna) [inline crates/apollo-compiler/src/database/hir_db.rs:25]:
```suggestion
    /// Return a `PrecomputedTypeSystem` which can be used to add a type system to a new instance of `ApolloCompiler`.
    fn precomputed_type_system(&self) -> Arc<PrecomputedTypeSystem>;
```

File: crates/apollo-compiler/src/database/hir.rs (hunk 3 above)
Comment (SimonSapin) [reply crates/apollo-compiler/src/database/hir.rs:33]: Having the struct for "type system definitions, supporting sources, and other stuff derived from definitions" be called just "type system" feels a bit weird, but I don’t really have a better suggestion

File: crates/apollo-compiler/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimonSapin) [reply crates/apollo-compiler/src/lib.rs:93]: I feel they’re so different except in form. One is a string while the other is structure data we got out of parsing that string, but they still represent the same thing

**Review by lrlna — APPROVED** [review]
i think this is good to go \o/
